azurelinuxagent/ga/exthandlers.py

```diff
@@ -949,31 +993,37 @@ def launch_command(self, cmd, timeout=300):
         begin_utc = datetime.datetime.utcnow()
         self.logger.verbose("Launch command: [{0}]", cmd)
         base_dir = self.get_base_dir()
+
+        def sanitize(s):
+            return ustr(s, encoding='utf-8', errors='backslashreplace')
+
         try:
-            devnull = open(os.devnull, 'w')
-            child = subprocess.Popen(base_dir + "/" + cmd,
-                                     shell=True,
-                                     cwd=base_dir,
-                                     stdout=devnull,
-                                     env=os.environ)
+            # This should be .run(), but due to the wide variety
+            # of Python versions we must support we must use .communicate().
+            process = subprocess.Popen(os.path.join(base_dir, cmd),
+                                  shell=True,
+                                  cwd=base_dir,
+                                  stdout=subprocess.PIPE,
+                                  stderr=subprocess.PIPE,
+                                  env=os.environ)
+            stdout, stderr = process.communicate(timeout)
+        except subprocess.TimeoutExpired:
+            process.kill()
+            stdout, stderr = process.communicate()
+            msg = format_stdout_stderr(sanitize(stdout), sanitize(stderr))
+            raise ExtensionError("Timeout({0}): {1}\n{2}".format(timeout, cmd, msg))
         except Exception as e:
-            #TODO do not catch all exception
+            process.kill()
```

Comment:
The previous code called os.kill(process.pid, 9) whereas this code calls process.kill(). I made this change because this is the way Python does it.  I can be convinced either way.